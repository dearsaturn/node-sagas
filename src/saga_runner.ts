import Bluebird from 'bluebird';
import {enums} from '@social-native/snpkg-snapi-authorization';

import {
    IAction,
    IBaseSagaContext,
    IEffectDescription,
    ArrayCombinator,
    RecordCombinator,
    Saga,
    Middleware,
    Next
} from './types';
import {ConsumerMessageBus} from './consumer_message_bus';
import {ProducerMessageBus} from './producer_message_bus';
import {
    isActionChannelEffectDescription,
    isTakeEffectDescription,
    isPutEffectDescription,
    isCallEffectDescription,
    isTakeActionChannelEffectDescription,
    isEffectCombinatorDescription,
    isDelayEffectDescription,
    isGenerator
} from './type_guard';

const {
    WORKER_USER_IDENTITY_HEADER: {WORKER_USER_ID, WORKER_USER_ROLES}
} = enums;

export class SagaRunner<InitialActionPayload, Context extends IBaseSagaContext> {
    protected runEffectWithMiddleware: <EffectDescription extends IEffectDescription>(
        effect: EffectDescription,
        context: Context
    ) => Promise<any>;

    constructor(
        private consumerMessageBus: ConsumerMessageBus,
        private producerMessageBus: ProducerMessageBus,
        middlewares: Array<Middleware<IEffectDescription, Context>> = []
    ) {
        const initialNext: Next<IEffectDescription, Context> = async (effect, ctx) => {
            return this.runEffect(effect, ctx);
        };

        this.runEffectWithMiddleware = middlewares
            .reduceRight((previousNext, middleware: Middleware<IEffectDescription, Context>) => {
                return middleware(previousNext);
            }, initialNext)
            .bind(this);
    }

    public runSaga = async (
        initialAction: IAction<InitialActionPayload>,
        context: Context,
        saga: Saga<InitialActionPayload, Context>
    ): Promise<any> => {
        this.consumerMessageBus.startTransaction(initialAction.transaction_id);
        const result = await this.runGeneratorFsm(saga(initialAction, context), context);
        this.consumerMessageBus.stopTransaction(initialAction.transaction_id);
        return result;
    };

    // tslint:disable-next-line: cyclomatic-complexity
    public runEffect = async <EffectDescription extends IEffectDescription>(
        effectDescription: EffectDescription,
        context: Context
    ) => {
        if (isEffectCombinatorDescription(effectDescription)) {
            const {effects, combinator} = effectDescription;

            if (Array.isArray(effects)) {
                const withRunningEffects: Array<Promise<any>> = effects.map(effect =>
                    this.runEffect(effect, context)
                );

                return await (combinator as ArrayCombinator<IAction>)(withRunningEffects);
            } else if (typeof effects === 'object') {
                const withRunningEffects: Record<string, Promise<any>> = Object.keys(
                    effects
                ).reduce((obj, key) => {
                    return {
                        ...obj,
                        [key]: this.runEffect(effects[key], context)
                    };
                }, {} as Record<string, Promise<any>>);

                return await (combinator as RecordCombinator<IAction>)(withRunningEffects);
            }

            throw new Error(
                'Incompatible effects passed into combinator. Must be an array or object of effects'
            );
        }

        if (isActionChannelEffectDescription(effectDescription)) {
            for (const topic of effectDescription.topics) {
                this.consumerMessageBus.registerTopicObserver({
                    transactionId: effectDescription.transactionId,
                    topic,
                    observer: effectDescription.observer
                });

                await this.consumerMessageBus.streamActionsFromTopic(topic);
            }

            return effectDescription;
        }

        // If this effect already has a stream buffer for events matching the pattern,
        // then just take from the buffer
        if (isTakeActionChannelEffectDescription(effectDescription)) {
            return await effectDescription.buffer.take();
        }

        if (isDelayEffectDescription(effectDescription)) {
            const {delayInMilliseconds, payload} = effectDescription;
            await Bluebird.delay(delayInMilliseconds);
            return payload;
        }

        if (isTakeEffectDescription(effectDescription)) {
            await Bluebird.map(effectDescription.topics, async topic => {
                this.consumerMessageBus.registerTopicObserver({
                    transactionId: effectDescription.transactionId,
                    topic,
                    observer: effectDescription.observer
                });

                await this.consumerMessageBus.streamActionsFromTopic(topic);
            });

            return await effectDescription.buffer.take();
        }

        if (isPutEffectDescription(effectDescription)) {
            const action: IAction<typeof effectDescription.payload> = {
                topic: effectDescription.pattern,
                transaction_id: effectDescription.transactionId,
                payload: effectDescription.payload
            };

            if (
                context.headers &&
                context.headers[WORKER_USER_ID] &&
                context.headers[WORKER_USER_ROLES]
            ) {
                action.userId = context.headers[WORKER_USER_ID];
                action.userRoles = context.headers[WORKER_USER_ROLES].split(',');
            }

            await this.producerMessageBus.putAction(action);

            return;
        }

        if (isCallEffectDescription(effectDescription)) {
            const result = await effectDescription.effect(...(effectDescription.args || []));

            if (isGenerator(result)) {
                return this.runGeneratorFsm(result, context);
            }

            return result;
        }
    };

    protected async runGeneratorFsm<Returned = any | undefined>(
        machine: Generator,
        context: Context,
        {
            previousGeneratorResponse = null,
            didThrow = false
        }: {
            previousGeneratorResponse: any;
            didThrow: boolean;
        } = {previousGeneratorResponse: null, didThrow: false}
    ): Promise<Returned> {
        /**
         * Dereferencing the receiver removes its context, so we need to bind it back to the machine.
         */
        const receiver = didThrow ? machine.throw.bind(machine) : machine.next.bind(machine);

        const {done, value} = receiver(previousGeneratorResponse) as IteratorResult<
            IEffectDescription,
            Returned
        >;

        if (done) {
            return value as Returned;
        }

        try {
            const result = await this.runEffectWithMiddleware(value as IEffectDescription, context);

            return this.runGeneratorFsm(machine, context, {
                previousGeneratorResponse: result,
                didThrow: false
            });
        } catch (error) {
            return this.runGeneratorFsm(machine, context, {
                previousGeneratorResponse: error,
                didThrow: true
            });
        }
    }
}
