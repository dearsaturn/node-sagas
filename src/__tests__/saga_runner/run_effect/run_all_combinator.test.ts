import {SagaRunner} from 'saga_runner';
import {withTopicCleanup} from '../../kafka_utils';
import {runnerUtilityFactory} from '../runner_utility_factory';
import {DEFAULT_TEST_TIMEOUT} from '../../constants';

describe(SagaRunner.name, function() {
    describe('all', function() {
        it(
            'runs all of an array of effects',
            async function() {
                await withTopicCleanup(['test-all-1'])(async ([topic]) => {
                    const {effectBuilder, runner, closeBuses} = await runnerUtilityFactory();
                    const channel = effectBuilder.actionChannel<{
                        bart_simpson: string;
                    }>(topic);
                    await runner.runEffect(channel);
                    await runner.runEffect(
                        effectBuilder.put(topic, {
                            bart_simpson: 'first'
                        })
                    );

                    await runner.runEffect(
                        effectBuilder.put(topic, {
                            bart_simpson: 'second'
                        })
                    );

                    const payload = await runner.runEffect(
                        effectBuilder.all([
                            effectBuilder.take(channel),
                            effectBuilder.take(channel)
                        ])
                    );

                    expect(payload).toMatchSnapshot();

                    await closeBuses();
                });
            },
            DEFAULT_TEST_TIMEOUT
        );

        it(
            'runs all of a record of effects',
            async function() {
                await withTopicCleanup(['test-all-1'])(async ([topic]) => {
                    const {effectBuilder, runner, closeBuses} = await runnerUtilityFactory();
                    const channel = effectBuilder.actionChannel<{
                        bart_simpson: string;
                    }>(topic);
                    await runner.runEffect(channel);
                    await runner.runEffect(
                        effectBuilder.put(topic, {
                            bart_simpson: 'first'
                        })
                    );

                    await runner.runEffect(
                        effectBuilder.put(topic, {
                            bart_simpson: 'second'
                        })
                    );

                    const payload = await runner.runEffect(
                        effectBuilder.all({
                            one: effectBuilder.take(channel),
                            two: effectBuilder.take(channel)
                        })
                    );

                    expect(payload).toMatchSnapshot();

                    await closeBuses();
                });
            },
            DEFAULT_TEST_TIMEOUT
        );
    });
});
