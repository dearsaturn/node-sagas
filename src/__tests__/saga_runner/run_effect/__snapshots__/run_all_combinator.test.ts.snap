// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaRunner all runs all of a record of effects 1`] = `
Object {
  "one": Object {
    "payload": Object {
      "bart_simpson": "first",
    },
    "topic": "test-all-1",
    "transaction_id": "static-transaction-id",
  },
  "two": Object {
    "payload": Object {
      "bart_simpson": "second",
    },
    "topic": "test-all-1",
    "transaction_id": "static-transaction-id",
  },
}
`;

exports[`SagaRunner all runs all of an array of effects 1`] = `
Array [
  Object {
    "payload": Object {
      "bart_simpson": "first",
    },
    "topic": "test-all-1",
    "transaction_id": "static-transaction-id",
  },
  Object {
    "payload": Object {
      "bart_simpson": "second",
    },
    "topic": "test-all-1",
    "transaction_id": "static-transaction-id",
  },
]
`;
